src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtAmazonDynamoDbStreamsBySharedTable.java

```diff
@@ -43,30 +44,108 @@ public GetShardIteratorResult getShardIterator(GetShardIteratorRequest request)
         return super.getShardIterator(request);
     }
 
-    // keeps going until it either reaches the end of stream or finds a record. This is so clients
-    // that don't cache shard iterators are guaranteed to make progress eventually.
+    /**
+     * Keeps fetching records until it reaches:
+     * <ol>
+     * <li>limit,</li>
+     * <li>current or absolute end of shard, or</li>
+     * <li>time bound</li>
+     * </ol>
+     * Returns a subtype of {@link GetRecordsResult} that includes the last consumed sequence number from the underlying
+     * stream shard, so that clients can request a new shard iterator where they left off.
+     */
     @Override
-    protected GetRecordsResult getRecords(GetRecordsRequest request, Function<Record, MtRecord> recordMapper,
-        Predicate<MtRecord> recordFilter) {
+    protected MtGetRecordsResult getRecords(GetRecordsRequest request, Function<Record, MtRecord> recordMapper,
+                                            Predicate<MtRecord> recordFilter) {
         Optional.ofNullable(request.getLimit()).ifPresent(limit -> checkArgument(limit > 0 && limit <= MAX_LIMIT));
 
-        List<Record> mtRecords = new ArrayList<>(Optional.ofNullable(request.getLimit()).orElse(MAX_LIMIT));
-        String iterator = request.getShardIterator();
+        final int limit = Optional.ofNullable(request.getLimit()).orElse(MAX_LIMIT);
+        final long timeLimit = mtDynamoDb.getGetRecordsTimeLimit();
+        final long time = mtDynamoDb.getClock().millis();
 
-        // keep fetching records until we reach the (current) end or find at least one record for this tenant
-        do {
-            GetRecordsResult result = dynamoDbStreams.getRecords(request.clone().withShardIterator(iterator));
-            iterator = result.getNextShardIterator();
-            List<Record> records = result.getRecords();
-            if (records.isEmpty()) {
-                break;
-            }
-            records.stream().map(recordMapper).filter(recordFilter).collect(toCollection(() -> mtRecords));
-        } while (mtRecords.isEmpty() && iterator != null);
+        final MtGetRecordsResult result = new MtGetRecordsResult()
+            .withRecords(new ArrayList<>(limit))
+            .withNextShardIterator(request.getShardIterator())
+            .withLastSequenceNumber(null);
```

Comment:
lastSequenceNumber by default will be null, is this line needed?
`.withLastSequenceNumber(null);`

Reply:
That's a good point. It's a remnant of a previous iteration where I had explicit variables for each and was creating the result after the loop. I'll take this line out.